src/filesystem.cpp

```diff
@@ -265,6 +277,191 @@ void FileAccess::closef()
     }
 }
 
+void FileAccess::asyncopfinished(void *param)
+{
+    AsyncIOContext *context = (AsyncIOContext *)param;
+    if (context->waiter)
+    {
+        context->waiter->notify();
+    }
+}
+
+AsyncIOContext *FileAccess::asyncfopen(string *f)
+{
+    localname.resize(1);
+    updatelocalname(f);
+
+    LOG_verbose << "Async open start";
+    AsyncIOContext *context = newasynccontext();
+    context->op = AsyncIOContext::OPEN;
+    context->access = AsyncIOContext::ACCESS_READ;
+
+    context->buffer = (byte *)f->data();
+    context->len = f->size();
+    context->waiter = waiter;
+    context->userCallback = asyncopfinished;
+    context->userData = context;
+    context->pos = size;
+    context->fa = this;
+
+    context->failed = !sysstat(&mtime, &size);
+    context->retry = this->retry;
+    context->finished = true;
+    context->userCallback(context->userData);
+    return context;
+}
+
+bool FileAccess::asyncopenf()
+{
+    if (!localname.size())
+    {
+        return true;
+    }
+
+    if (numops)
+    {
+        numops++;
+        return true;
+    }
+
+    m_time_t curr_mtime;
+    m_off_t curr_size;
+
+    if (!sysstat(&curr_mtime, &curr_size) || curr_mtime != mtime || curr_size != size)
+    {
+        return false;
+    }
+
+    LOG_debug << "Opening async file handle for reading";
+    bool result = sysopen(true);
+    if (result)
+    {
+        numops++;
+    }
+    else
+    {
+        LOG_err << "Error opening async file handle";
+    }
+    return result;
+}
+
+void FileAccess::asyncclosef()
+{
+    numops--;
```

Comment:
It should never happen. `asyncopenf` should be always used at the beginning of an async operation, and `asyncclosef` at the end. If that doesn't happen, we would have a bug.

If we add that check, we could just call `sysclose` more times and we would also make the bug more difficult to detect, because otherwise it should be detected in the destructor of `FileAccess` with the assert:
`assert(!numops);`